### PR TITLE
socket-util: tighten CMSG_TYPED_DATA() alignment checks

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

**Comment by poettering** [comment]
(this PR is not complete yet, just looking for CI results for now. Will push a new version shortly, that ports more stuff over to CMSG_TYPED_DATA())

**Review by yuwata — APPROVED** [review]
LGTM.

File: src/basic/socket-util.h [hunk 1]
```diff
@@ -175,9 +175,16 @@ int flush_accept(int fd);
...
+/* Returns the cmsghdr's data pointer, but safely cast to the specified type. Does two alignment checks: one
+ * at compile time, that the requested type has a smaller or same alignment as 'struct cmsghdr', and one
+ * during runtime, that the actual pointer matches the alignment too. This is supposed to catch cases such as
+ * 'struct timeval' is embedded into 'struct cmsghdr' on architectures where the alignment of the former is 8
+ * bytes (because of a 64bit time_t), but of the latter is 4 bytes (because size_t is 32bit), such as
+ * riscv32. */
 #define CMSG_TYPED_DATA(cmsg, type)                                     \
         ({                                                              \
                 struct cmsghdr *_cmsg = cmsg;                           \
+                assert_cc(__alignof__(type) <= __alignof__(struct cmsghdr)); \
```
Comment (medhefgo) [inline src/basic/socket-util.h:187]: Why not use `stdalign.h`?

**Review by bluca — APPROVED** [review]

**Comment by yuwata** [comment]
Let's wait for @kraj's response.

**Comment by kraj** [comment]
with this changeset applied I get build time errors as expected I guess
```
 ../git/src/libsystemd-network/icmp6-util.c:204:73: error: static assertion failed due to requirement '__alignof(struct timeval) <= __alignof(struct cmsghdr)': __alignof__(struct timeval) <= __alignof__(struct cms
  ghdr)
                          triple_timestamp_from_realtime(&t, timeval_load(CMSG_TYPED_DATA(cmsg, struct timeval)));
                                                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  ../git/src/basic/socket-util.h:187:17: note: expanded from macro 'CMSG_TYPED_DATA'
                  assert_cc(__alignof__(type) <= __alignof__(struct cmsghdr)); \
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  ../git/src/fundamental/macro-fundamental.h:108:25: note: expanded from macro 'assert_cc'
  #define assert_cc(expr) static_assert(expr, #expr)
                          ^             ~~~~
  /mnt/b/yoe/master/build/tmp/work/riscv32-yoe-linux/systemd/1_253.1-r0/recipe-sysroot/usr/include/assert.h:144:24: note: expanded from macro 'static_assert'
  # define static_assert _Static_assert
                         ^
  ../git/src/libsystemd-network/icmp6-util.c:204:73: note: expression evaluates to '8 <= 4'
                          triple_timestamp_from_realtime(&t, timeval_load(CMSG_TYPED_DATA(cmsg, struct timeval)));
                                                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  ../git/src/basic/socket-util.h:187:45: note: expanded from macro 'CMSG_TYPED_DATA'
                  assert_cc(__alignof__(type) <= __alignof__(struct cmsghdr)); \
                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  ../git/src/fundamental/macro-fundamental.h:108:39: note: expanded from macro 'assert_cc'
  #define assert_cc(expr) static_assert(expr, #expr)
                                        ^~~~
  1 error generated.
```

The full compile log for rv32 is here http://sprunge.us/80qGf1

**Comment by yuwata** [comment]
@kraj Thanks.
Merging. I will address @medhefgo's comment later in a different PR.

**Pull request merged**